### PR TITLE
GuidedTours: Frontend preview

### LINKED ISSUE #4380: GuidesTours: Preview frontend
_I want to see my site, so that I get immediate feedback on what I’ve just chosen._

**TODO:**
- [ ] Use WebPreview, modified to ‘fake’ frontend
- [ ] Show on tour start
- [ ] Redirect/hide preview after first step? Maybe redirect to blogs/pages view

**UI:**
screen shot 2016-03-29 at 13 38 09

### PR DESCRIPTION
**Pull request by ehg**
_edits by @mcsf_

~~Based off #4391~~
Intends to close #4380

### TODO
- [x] Might be worth looking into `DesignPreview` for this.
- [x] Properly hide/disable `primary` when showing preview
- [x] Properly position preview relative to sidebar (currently breaks with resizes)
- [ ] z-index: Position preview below `.environment-badge`
- [x] Be more defensive, codewise, when no site selected
- [x] Hide the sidebar at first

### Current state
calypso-guidestours-160408
(Previously: [#](https://cloud.githubusercontent.com/assets/215074/14205997/bbeffee6-f807-11e5-9c1d-2997f5c79bdb.gif))

### To test
- Add `?tour=main` to any page **where a site has been selected**
- Reload the page, hopefully see a preview and a tour
- Play around a bit

### CODE REVIEW

**Comment by mcsf** [comment]
> Might be worth looking into `DesignPreview` for this.

Given that `DesignPreview` hasn't landed yet and that judging from #4140 it seems like it does a whole lot more than just what we need, we should just merge this the way it's done with `WebPreview` and `?iframe=true&preview=true`.

Looping @sirbrillig, who might be interested in this use case.

**Comment by ehg** [comment]
> we should just merge this the way it's done with WebPreview and ?iframe=true&preview=true.

Yeah, I had a closer look, and we should be OK with our own approach for the MVP. We could then maybe use the cool `layoutFocus` approach later.

File: client/components/main/index.jsx [hunk 1]
```diff
@@ -17,3 +29,44 @@ module.exports = React.createClass( {
 		);
 	}
 } );
+
+/**
+ * As apparent below, the idea is simple: connect Main to Redux so it knows
+ * whether it should render a guided tour rather than `primary`.
+ *
```
Comment (ehg) [inline client/components/main/index.jsx]: Why should `Main` care about this? Aren't we just overlaying the preview?

Comment (mcsf) [inline client/components/main/index.jsx]: Right, I should've made the motivation clear:

The preview is currently overlaid, but if you try resizing you'll see a lot of funkiness going on, as the sidebar shrinks, then hides. Specifically, you'll see a glimpse of `primary` in between the sidebar and the overlay.

My reasoning, which I should've shared yesterday, is that it's probably a bad idea to rely on a `position: fixed` element to make sure we fully hide Calypso consistently across different viewports and wild scenarios. Thus, I'd like Main — since it hierarchically sits pretty close to `primary` (though sadly not _above_ it; suggestions for other components that would work across sections are welcome)  — to do the following, roughly speaking:

```js
const classes = classnames( this.props.className, 'main', {
    // I'm terrible at CSS-oriented naming
    'main__is-guidestour-preview-active': shouldShowGuides
} );
return (
  <main className={ classes } role="main">
    { shouldShowGuides && <WebPreview { ...previewProps } /> }
    { this.props.children }
  </main>
);
```

That would:

1. Insert the preview at the same document level of the content / `primary`, thus gaining more flexibility for good overlay positioning.
1. Mark `<main>` with a stateful class so as to make it easier to use plain CSS to hide a Calypso section while the preview is showing.

Is this sound? :)

Comment (mcsf) [inline client/components/main/index.jsx]: A chat with @ehg lead to the quick conclusion that, while the general idea likely makes sense, `client/layout` is a much more suitable component than `Main`, so I'll get to that.

**Comment by sirbrillig** [comment]
Yeah, I would not rely on `DesignPreview` yet, although it could definitely be a useful tool for this situation. Things are still too much in flux there. Very cool, though. Thanks for the ping!

**Comment by mcsf** [comment]
@ehg @lsinger — 9a0d87d is a pretty big and hideous beast, but it switches from touching Main to touching Layout, and implements a bunch of other stuff (handling viewport resizes and also guessing the sidebar width while the UI is still being built for the first time; piping more things through Redux, e.g. killing `state.currentStep`, etc.).

**Comment by ehg** [comment]
Cool.

According to the mockups, we don't actually need to show the sidebar…

screen shot 2016-04-06 at 20 29 10

**Comment by mcsf** [comment]
Good point. So the preview could be initialized with `left: 0` and only once My Sites is clicked would the preview make room for the sidebar.

**Comment by ehg** [comment]
>  and only once My Sites is clicked would the preview make room for the sidebar.

Yeah, although I'm kind of wary about leaving the preview there, since there's no current way to get back to this particular form of preview. I think it'll be more harmonious when we have Preview Everywhere, but we could hide the preview once My Sites is clicked now. /cc @shaunandrews

**Comment by mcsf** [comment]
> Yeah, although I'm kind of wary about leaving the preview there, since there's no current way to get back to this particular form of preview.

Sorry, I didn't mean closing the preview, I meant, literally, to _make room_ for the sidebar by resizing the preview component so that we can see both.

**Comment by ehg** [comment]
> Sorry, I didn't mean closing the preview, I meant, literally, to make room for the sidebar by resizing the preview component so that we can see both.

How about we don't ever show both? :)

**Comment by ehg** [comment]
Flipping this over to design review, so we can nail the UX.

**Comment by folletto** [comment]
The way I could see this working would be:

1. When `?tour=main` is triggered, should instantly start with the preview visible (no fade-in animation) and the masterbar at the top.
2. All the steps in the sequence will keep the site preview on.
3. I'd however complete, iterate and gather feedback on the script first. Guided tours text/flow/branching is the most important bit to nail here.

That said, I'm still a bit not comfortable in showing a UI that is never going to be replicated, so my advice would be to wait for `layoutPreview` to land, and then review this, so moving this part of the UI to a different milestone (M2?).

**Comment by mcsf** [comment]
Big painful rebase, following #4468, pushed. Now to implement UX changes.

**Comment by mcsf** [comment]
Experimenting with a full-width preview shown at the start of the tour. Screencast:

calypso-guidestours-160408

**Comment by mcsf** [comment]
Aside from needing a design review, this is good for a code review. (I'll remove the commented-out stuff and squash the commits post-review.)

**Comment by lsinger** [comment]
Awesome! 

I usually test by going to e.g. /me?tour-main, and in this case the tour never shows up. Do we need to address that? We're making sure people only land on the proper URLs though, right?

**Comment by mcsf** [comment]
That was intentional, as we need some site to be selected. If we'd
like to have the tour operating from /me, we need the site info to
e.g. be injected into the GuidesTours state.

**Comment by lsinger** [comment]
> That was intentional, as we need some site to be selected. If we'd like to have the tour operating from /me, we need the site info to e.g. be injected into the GuidesTours state.

It would be nice to have it as resilient as possible. Could it make sense (and would it be relatively low effort) to just use the most recently created site (from the set of the user's sites) if none is given?

File: client/components/main/index.jsx [hunk 2]
```diff
@@ -3,17 +3,17 @@
 /**
```
Comment (ehg) [inline client/components/main/index.jsx]: Could this go in a separate PR? :P

**Comment by ehg** [comment]
> Experimenting with a full-width preview shown at the start of the tour.

I think the design consensus is to show the sidebar over the preview, when it's first opened via a section. Maybe we could get away with manipulating the z-index here? Ref: https://github.com/Automattic/wp-calypso/pull/4461#issuecomment-206830757 & https://marvelapp.com/c1che6#11469689

File: client/layout/index.jsx [hunk 3]
```diff
@@ -170,6 +173,8 @@ Layout = React.createClass( {
 					{ this.renderWelcome() }
 					{ this.renderEmailVerificationNotice() }
 					<GlobalNotices id="notices" notices={ notices.list } forcePinned={ 'post' === this.props.section.name } />
+					<GuidesPreview showPreview={ this.props.shouldShowGuidesPreview }
```
Comment (ehg) [inline client/layout/index.jsx]: Any reason this has to go in `Layout`?

**Comment by shaunandrews** [comment]
I dig the direction this is heading. The preview takes a few seconds to load, and you're seeing something like this:

screen shot 2016-04-11 at 11 29 05 am

It'd be cool if we had something better to show there. Maybe a blurred out, low-res image of their website. Since this is focused on new users (and we may not have an mShot to use) maybe we could use the theme preview image, at least for Headstart'd sites.? Something like this:

artboard

**Comment by folletto** [comment]
> maybe we could use the theme preview image, at least for Headstart'd sites

Brilliant idea. :100:

**Comment by shaunandrews** [comment]
Fun bug where the preview kept reloading new instances within itself:

screen shot 2016-04-12 at 10 17 26 am

screen shot 2016-04-12 at 10 18 16 am

Maybe related to the private browser window?

File: client/layout/index.jsx (hunk 3 above)
Comment (mcsf) [inline client/layout/index.jsx]: IIRC it was motivated by the possibility of moving away from fixed positioning because of how it interacted with the sidebar and primary. It doesn't matter as much now.

**Comment by ehg** [comment]
Closing this for now, in favour of https://github.com/Automattic/wp-calypso/pull/5117.
